### PR TITLE
Add deepcopy functionality to parametrized modules

### LINKED ISSUE #69413: [bug] `parametrize` can't be used together with `swa_utils.AveragedModel`
## 🐛 Bug
The `deepcopy` used in `AveragedModel.__init__` makes it incompatible with `parametrization`.

## Scripts to Reproduce
```python3
import torch
import torch.nn as nn

from torch.optim.swa_utils import AveragedModel
from torch.nn.utils import parametrize

if __name__ == '__main__':
    model = nn.Linear(3, 4)
    b = AveragedModel(model)

    class Param(nn.Module):
        def forward(self, X: torch.Tensor) -> torch.Tensor:
            return X.clamp(0.0, 1.0)

    parametrize.register_parametrization(model, 'weight', Param())

    c = AveragedModel(model)
```

## Output

```
  File "C:\Users\■\miniconda3\lib\site-packages\torch\optim\swa_utils.py", line 89, in __init__
    self.module = deepcopy(model)
  File "C:\Users\■\miniconda3\lib\copy.py", line 161, in deepcopy
    rv = reductor(4)
  File "C:\Users\■\miniconda3\lib\site-packages\torch\nn\utils\parametrize.py", line 285, in getstate
    raise RuntimeError(
RuntimeError: Serialization of parametrized modules is only supported through state_dict(). See:
https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-loading-a-general-checkpoint-for-inference-and-or-resuming-training
```

## Environment
- python 3.9
- pytorch 1.10

### PR DESCRIPTION
**Pull request by n-zhuravlev**
Fixes #69413

After applying parametrization to any `nn.Module` we lose the ability to create a deepcopy of it e.g. it makes it impossible to wrap a module by an `AveragedModel`.
Specifically, the problem is that the `deepcopy` tries to invoke `__getstate__` if object hasn't implemented its own `__deepcopy__` magic method. But we don't allow serialization of the parametrized modules: `__getstate__` raises an error. 
My solution is just to create a default `__deepcopy__` method when it doesn't exist yet.

### CODE REVIEW

File: torch/nn/utils/parametrize.py
Comment (Lezcano) [inline torch/nn/utils/parametrize.py:289]: Could you split this into two assignments for clarity?

File: test/test_nn.py
Comment (Lezcano) [inline test/test_nn.py:3327]: After this, can you register the paramtrisation also on `bias` just to make sure that this works on objects with two parametrisations (I don't see any reason why it shouldn't), and then also register `AddOne` again on `weight` to see that it works on objects with two parametrisations on the same tensor?

**Review by lezcano — COMMENTED** [review]
This looks very good, thanks! I left a couple comments.

I'll let @albanD have the last say on this one, but from my side it looks good :)

File: torch/nn/utils/parametrize.py
Comment (n-zhuravlev) [reply torch/nn/utils/parametrize.py]: Sure! Done :heavy_check_mark:

File: test/test_nn.py
Comment (n-zhuravlev) [reply test/test_nn.py]: Yes, I agree with you. Had to rewrite a bit, not to make code messy.
To make sure we are on the same page:
For both models (with `__deepcopy__` implementation and without):
- Checking that it works for pzn registered for the `weight`
- Checking that it works for pzn registered for the `weight` and `bias`
- Checking that it works for pzn registered for the `bias` and several pzn registered for the `weight` 

Take a look once again please!

File: torch/nn/utils/parametrize.py
Comment (albanD) [inline torch/nn/utils/parametrize.py:288]: not needed

File: torch/nn/utils/parametrize.py
Comment (albanD) [inline torch/nn/utils/parametrize.py:289]: You should first check if self is not in the memo.

File: torch/nn/utils/parametrize.py
Comment (albanD) [inline torch/nn/utils/parametrize.py:291]: You can do `replica.__dict__ = deepcopy(self.__dict__)`
You need also to handle slots by doing something like:
```python
  slots_to_save = copyreg._slotnames(self.__class__)  # type: ignore[attr-defined]
  for slot in slots_to_save:
      if hasattr(self, slot):
          setattr(replica, slot, deepcopy(getattr(self, slot), memo))
```

File: test/test_nn.py
Comment (albanD) [inline test/test_nn.py:3300]: I don't think this is a very accurate emulation. For testing purposes, you can do a dummy function here that just returns a new Module.

File: test/test_nn.py
Comment (albanD) [inline test/test_nn.py:3314]: self.assertIsNot

File: test/test_nn.py
Comment (albanD) [inline test/test_nn.py:3313]: Note that the dict does NOT contain the parameters/biases. So this only checks for the other attributes.
If you want to check that all the parameters and buffers are here, you can check the result of `.named_parameters()` and `named_buffers()`.

**Review by albanD — COMMENTED** [review]
Some small comments, but looks ok overall.

File: torch/nn/utils/parametrize.py
Comment (n-zhuravlev) [reply torch/nn/utils/parametrize.py]: Sure, done :heavy_check_mark: 
The only thing here is that now, when method accepts `memo` without a default value `mypy` prints an error for the method assignment `dct["__deepcopy__"] = default_deepcopy` :
```
error: Incompatible types in assignment (expression has type "Callable[[Any, Any], Any]", target has type "Callable[[Any], Any]")  [assignment]
```
Just added type check suppression
```python
dct["__deepcopy__"] = default_deepcopy  # type: ignore[assignment]
```

File: torch/nn/utils/parametrize.py
Comment (n-zhuravlev) [reply torch/nn/utils/parametrize.py]: Yes, now deepcopy is created only when `id(self) not in memo`

File: test/test_nn.py
Comment (n-zhuravlev) [reply test/test_nn.py]: Changed to `self.assertIsNot` :heavy_check_mark:

File: test/test_nn.py
Comment (n-zhuravlev) [reply test/test_nn.py:3300]: Yes, it's not an accurate deepcopy, but if I create a new Module like `self.__class__()` I should register all parametrizations once again for this module depending on ones that have already registered for `self`. 
As this simple module contains all weights, biases and attributes in `__dict__` don't you think that it's enough just to create a `deepcopy` of it's `__dict__`?

File: test/test_nn.py
Comment (n-zhuravlev) [reply test/test_nn.py:3313]: This module contains only `weight` and `bias` parameters and they are extracted (their existence is checked) above. Just wanted to check all other names in `__dict__` here

File: torch/nn/utils/parametrize.py
Comment (n-zhuravlev) [reply torch/nn/utils/parametrize.py]: 1. Sure, replaced to `replica.__dict__ = deepcopy(self.__dict__, memo)` :heavy_check_mark: 
2. Added code of slots copying from `Tensor.__deepcopy__`.

The only question I have here is about slots registering. I thought that our modules isn't assumed to use `__slots__` and just rely on regular `__dict__` attributes storing. To explain what I mean here's a toy example:

```python
class Model(nn.Module):
    
    __slots__ = ["weight"]
    
    def __init__(self):
        super().__init__()
        self.weight = nn.Parameter(torch.tensor([1., 1., 1., 1.]), requires_grad=True)

model = Model()
model_copy = deepcopy(model)

```
This code raises an error: when `deepcopy` tries to restore object via `__setstate__` it can't do it because it's assumed that all our attributes are stored inside the `__dict__`.

I will be very grateful if you clarify the situation for me.

File: torch/nn/utils/parametrize.py
Comment (Lezcano) [inline torch/nn/utils/parametrize.py:290]: Nit. We may want to go with the more efficient:
```python
obj = memo.get(id(self))
if obj is not None:
  return obj
```

File: test/test_nn.py
Comment (albanD) [reply test/test_nn.py:3300]: Ok, We can keep it then.

File: test/test_nn.py
Comment (albanD) [reply test/test_nn.py:3313]: Ok then.

**Review by albanD — COMMENTED** [review]
Looks good to me.
Just a small nit.

File: torch/nn/utils/parametrize.py
Comment (albanD) [reply torch/nn/utils/parametrize.py]: I don't think parameters or buffers can be slots indeed. But users might use other attributes as slots.
Note that we are considering having some of the builtin attributes as slots as well.

Comment (n-zhuravlev) [reply torch/nn/utils/parametrize.py]: Got it, thank you!

File: torch/nn/utils/parametrize.py
Comment (n-zhuravlev) [reply torch/nn/utils/parametrize.py]: Agree, changed to proposed code to remove indentation :heavy_check_mark:

**Comment by n-zhuravlev** [comment]
So, what's next?
I see some failed CI tests but their error messages are not obvious for me, it needs some time to sort out the problem.
Can you give a small hint, whether this caused by the PR or I just should wait for more reviews?

File: torch/nn/utils/parametrize.py
Comment (pearu) [inline torch/nn/utils/parametrize.py:287]: Just a nitpick, feel free to ignore: It is not really a "default deepcopy" as the default deepcopy would fail due to the lack of serialization support.

**Review by pearu — APPROVED** [review]
I have a minor nit but overall LGTM!

File: torch/nn/utils/parametrize.py
Comment (albanD) [reply torch/nn/utils/parametrize.py:289]: Yes unspecified is better, but this should never be `None` in any case unless something goes very wrong

**Review by albanD — APPROVED** [review]

**Comment by n-zhuravlev** [comment]
:confused: Rebased on top of `viable/strict` but it has accidentally triggered another reviewers. @Lezcano can I exclude them?

**Comment by Lezcano** [comment]
The XLA failure seems unrelated. Merging.

**Pull request closed without merging**
